system/Database/BaseConnection.php

```diff
@@ -342,6 +342,13 @@ public function __construct(array $params)
         if (class_exists($queryClass)) {
             $this->queryClass = $queryClass;
         }
+
+        if ($this->failover !== []) {
+            // If there is a failover database, connect now to do failover.
+            // Otherwise, Query Builder creates SQL statement with the main database config
+            // (DBPrefix) even when the main database is down.
+            $this->initialize();
```

Comment:
Is there any reason why we wouldn't always initialize now, for consistency? This seems like it could introduce some strange varying behavior based on whether failover is configured or not.

Reply:
Because this is the minimum change to fix the bug.

Do you accept that removing the `if` and always initialize in the constructor?
It changes the current behavior completely, so it seems a breaking change.

Reply:
It seems calling `initialize()` just before running query came from CI3.

But CI4 has a lot of code like this:
```php
        if (empty($this->connID)) {
            $this->initialize();
        }
```

If we call`initialize()` always in the constructor, we could remove the code.

But the following test code will be fail:
```php
        $db = new MockConnection($this->options);
        $db->shouldReturn('connect', 123)->initialize();
```

Reply:
No problem. 
Personally I think calling initialize() always in the constructor is better and cleaner code.